From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Abhay Singh Gill <abhaygill017@gmail.com>
Date: Sat, 31 May 2025 07:17:21 +0530
Subject: [PATCH 18/31] base: Add the ability to generate key pairs for keybox
 spoofing

The code is used from :-
https://github.com/5ec1cff/TrickyStore

Needs :-
https://github.com/HELLBOY017/external_bouncycastle/commit/f70ea016239bccf2d3f4b60b79a7165ef3d047be

This allows us to enjoy spoofing keyboxes the right way even with broken TEE.

Signed-off-by: Abhay Singh Gill <abhaygill017@gmail.com>
Signed-off-by: Ritesh <SahanyRitesh@proton.me>
---
 .../internal/util/KeyboxChainGenerator.java   | 441 ++++++++++++++++++
 .../internal/util/KeyboxImitationHooks.java   | 309 ++++++------
 .../android/internal/util/KeyboxUtils.java    | 134 ++++++
 keystore/java/android/security/KeyStore2.java |   8 +-
 .../security/KeyStoreSecurityLevel.java       |  20 +
 .../AndroidKeyStoreKeyPairGeneratorSpi.java   |  24 +-
 6 files changed, 767 insertions(+), 169 deletions(-)
 create mode 100644 core/java/com/android/internal/util/KeyboxChainGenerator.java
 create mode 100644 core/java/com/android/internal/util/KeyboxUtils.java

diff --git a/core/java/com/android/internal/util/KeyboxChainGenerator.java b/core/java/com/android/internal/util/KeyboxChainGenerator.java
new file mode 100644
index 0000000000..b157cb6a15
--- /dev/null
+++ b/core/java/com/android/internal/util/KeyboxChainGenerator.java
@@ -0,0 +1,441 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Neoteric OS
+ * SPDX-FileCopyrightText: 2025 The Clover Project
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.android.internal.util;
+
+import android.app.ActivityThread;
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.content.pm.Signature;
+import android.hardware.security.keymint.Algorithm;
+import android.hardware.security.keymint.EcCurve;
+import android.hardware.security.keymint.KeyParameter;
+import android.hardware.security.keymint.Tag;
+import android.os.Binder;
+import android.os.Build;
+import android.security.keystore.KeyProperties;
+import android.system.keystore2.KeyDescriptor;
+import android.util.Log;
+
+import androidx.annotation.Nullable;
+
+import com.android.internal.org.bouncycastle.asn1.ASN1Boolean;
+import com.android.internal.org.bouncycastle.asn1.ASN1Encodable;
+import com.android.internal.org.bouncycastle.asn1.ASN1Enumerated;
+import com.android.internal.org.bouncycastle.asn1.ASN1Integer;
+import com.android.internal.org.bouncycastle.asn1.ASN1ObjectIdentifier;
+import com.android.internal.org.bouncycastle.asn1.ASN1OctetString;
+import com.android.internal.org.bouncycastle.asn1.ASN1Sequence;
+import com.android.internal.org.bouncycastle.asn1.DERNull;
+import com.android.internal.org.bouncycastle.asn1.DEROctetString;
+import com.android.internal.org.bouncycastle.asn1.DERSequence;
+import com.android.internal.org.bouncycastle.asn1.DERSet;
+import com.android.internal.org.bouncycastle.asn1.DERTaggedObject;
+import com.android.internal.org.bouncycastle.asn1.x500.X500Name;
+import com.android.internal.org.bouncycastle.asn1.x509.Extension;
+import com.android.internal.org.bouncycastle.asn1.x509.KeyUsage;
+import com.android.internal.org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import com.android.internal.org.bouncycastle.asn1.x509.Time;
+import com.android.internal.org.bouncycastle.cert.X509CertificateHolder;
+import com.android.internal.org.bouncycastle.cert.X509v3CertificateBuilder;
+import com.android.internal.org.bouncycastle.jce.provider.BouncyCastleProvider;
+import com.android.internal.org.bouncycastle.operator.ContentSigner;
+import com.android.internal.org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.security.Security;
+import java.security.cert.Certificate;
+import java.security.spec.ECGenParameterSpec;
+import java.security.spec.RSAKeyGenParameterSpec;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import javax.security.auth.x500.X500Principal;
+
+/**
+ * @hide
+ */
+public final class KeyboxChainGenerator {
+
+    private static final String TAG = "KeyboxChainGenerator";
+    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
+
+    private static final int ATTESTATION_APPLICATION_ID_PACKAGE_INFOS_INDEX = 0;
+    private static final int ATTESTATION_APPLICATION_ID_SIGNATURE_DIGESTS_INDEX = 1;
+    private static final int ATTESTATION_PACKAGE_INFO_PACKAGE_NAME_INDEX = 0;
+    private static final int ATTESTATION_PACKAGE_INFO_VERSION_INDEX = 1;
+
+    public static List<Certificate> generateCertChain(int uid, KeyDescriptor descriptor, KeyGenParameters params) {
+        dlog("Requested KeyPair with alias: " + descriptor.alias);
+        int size = params.keySize;
+        KeyPair kp;
+        try {
+            if (Objects.equals(params.algorithm, Algorithm.EC)) {
+                dlog("Generating EC keypair of size " + size);
+                kp = buildECKeyPair(params);
+            } else if (Objects.equals(params.algorithm, Algorithm.RSA)) {
+                dlog("Generating RSA keypair of size " + size);
+                kp = buildRSAKeyPair(params);
+            } else {
+                dlog("Unsupported algorithm");
+                return null;
+            }
+
+            X509v3CertificateBuilder certBuilder = new X509v3CertificateBuilder(
+                    KeyboxUtils.getCertificateHolder(
+                            Objects.equals(params.algorithm, Algorithm.EC)
+                                    ? KeyProperties.KEY_ALGORITHM_EC
+                                    : KeyProperties.KEY_ALGORITHM_RSA
+                    ).getSubject(),
+                    params.certificateSerial,
+                    new Time(params.certificateNotBefore),
+                    new Time(params.certificateNotAfter),
+                    params.certificateSubject,
+                    SubjectPublicKeyInfo.getInstance(
+                            ASN1Sequence.getInstance(kp.getPublic().getEncoded())
+                    )
+            );
+
+            KeyUsage keyUsage = new KeyUsage(KeyUsage.keyCertSign);
+            certBuilder.addExtension(Extension.keyUsage, true, keyUsage);
+            certBuilder.addExtension(createExtension(params, uid));
+
+            ContentSigner contentSigner;
+            if (Objects.equals(params.algorithm, Algorithm.EC)) {
+                contentSigner = new JcaContentSignerBuilder("SHA256withECDSA").build(KeyboxUtils.getPrivateKey(KeyProperties.KEY_ALGORITHM_EC));
+            } else {
+                contentSigner = new JcaContentSignerBuilder("SHA256withRSA").build(KeyboxUtils.getPrivateKey(KeyProperties.KEY_ALGORITHM_RSA));
+            }
+            X509CertificateHolder certHolder = certBuilder.build(contentSigner);
+            Certificate leaf = KeyboxUtils.getCertificateFromHolder(certHolder);
+            List<Certificate> chain = KeyboxUtils.getCertificateChain(leaf.getPublicKey().getAlgorithm());
+            chain.add(0, leaf);
+            dlog("Successfully generated X500 Cert for alias: " + descriptor.alias);
+            return chain;
+        } catch (Throwable t) {
+            Log.e(TAG, Log.getStackTraceString(t));
+        }
+        return null;
+    }
+
+    private static ASN1Encodable[] fromIntList(List<Integer> list) {
+        ASN1Encodable[] result = new ASN1Encodable[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            result[i] = new ASN1Integer(list.get(i));
+        }
+        return result;
+    }
+
+    private static Extension createExtension(KeyGenParameters params, int uid) {
+        try {
+            SecureRandom random = new SecureRandom();
+
+            byte[] bytes1 = new byte[32];
+            byte[] bytes2 = new byte[32];
+
+            random.nextBytes(bytes1);
+            random.nextBytes(bytes2);
+
+            ASN1Encodable[] rootOfTrustEncodables = {new DEROctetString(bytes1), ASN1Boolean.TRUE,
+                    new ASN1Enumerated(0), new DEROctetString(bytes2)};
+
+            ASN1Sequence rootOfTrustSeq = new DERSequence(rootOfTrustEncodables);
+
+            var Apurpose = new DERSet(fromIntList(params.purpose));
+            var Aalgorithm = new ASN1Integer(params.algorithm);
+            var AkeySize = new ASN1Integer(params.keySize);
+            var Adigest = new DERSet(fromIntList(params.digest));
+            var AecCurve = new ASN1Integer(params.ecCurve);
+            var AnoAuthRequired = DERNull.INSTANCE;
+
+            // To be loaded
+            var AosVersion = new ASN1Integer(getOsVersion());
+            var AosPatchLevel = new ASN1Integer(getPatchLevel());
+
+            var AapplicationID = createApplicationId(uid);
+            var AbootPatchlevel = new ASN1Integer(getPatchLevelLong());
+            var AvendorPatchLevel = new ASN1Integer(getPatchLevelLong());
+
+            var AcreationDateTime = new ASN1Integer(System.currentTimeMillis());
+            var Aorigin = new ASN1Integer(0);
+
+            var purpose = new DERTaggedObject(true, 1, Apurpose);
+            var algorithm = new DERTaggedObject(true, 2, Aalgorithm);
+            var keySize = new DERTaggedObject(true, 3, AkeySize);
+            var digest = new DERTaggedObject(true, 5, Adigest);
+            var ecCurve = new DERTaggedObject(true, 10, AecCurve);
+            var noAuthRequired = new DERTaggedObject(true, 503, AnoAuthRequired);
+            var creationDateTime = new DERTaggedObject(true, 701, AcreationDateTime);
+            var origin = new DERTaggedObject(true, 702, Aorigin);
+            var rootOfTrust = new DERTaggedObject(true, 704, rootOfTrustSeq);
+            var osVersion = new DERTaggedObject(true, 705, AosVersion);
+            var osPatchLevel = new DERTaggedObject(true, 706, AosPatchLevel);
+            var applicationID = new DERTaggedObject(true, 709, AapplicationID);
+            var vendorPatchLevel = new DERTaggedObject(true, 718, AvendorPatchLevel);
+            var bootPatchLevel = new DERTaggedObject(true, 719, AbootPatchlevel);
+
+            ASN1Encodable[] teeEnforcedEncodables;
+
+            // Support device properties attestation
+            if (params.brand != null) {
+                var Abrand = new DEROctetString(params.brand);
+                var Adevice = new DEROctetString(params.device);
+                var Aproduct = new DEROctetString(params.product);
+                var Amanufacturer = new DEROctetString(params.manufacturer);
+                var Amodel = new DEROctetString(params.model);
+                var brand = new DERTaggedObject(true, 710, Abrand);
+                var device = new DERTaggedObject(true, 711, Adevice);
+                var product = new DERTaggedObject(true, 712, Aproduct);
+                var manufacturer = new DERTaggedObject(true, 716, Amanufacturer);
+                var model = new DERTaggedObject(true, 717, Amodel);
+
+                teeEnforcedEncodables = new ASN1Encodable[]{purpose, algorithm, keySize, digest, ecCurve,
+                        noAuthRequired, origin, rootOfTrust, osVersion, osPatchLevel, vendorPatchLevel,
+                        bootPatchLevel, brand, device, product, manufacturer, model};
+            } else {
+                teeEnforcedEncodables = new ASN1Encodable[]{purpose, algorithm, keySize, digest, ecCurve,
+                        noAuthRequired, origin, rootOfTrust, osVersion, osPatchLevel, vendorPatchLevel,
+                        bootPatchLevel};
+            }
+
+            ASN1Encodable[] softwareEnforced = {applicationID, creationDateTime};
+
+            ASN1OctetString keyDescriptionOctetStr = getAsn1OctetString(teeEnforcedEncodables, softwareEnforced, params);
+
+            return new Extension(new ASN1ObjectIdentifier("1.3.6.1.4.1.11129.2.1.17"), false, keyDescriptionOctetStr);
+        } catch (Throwable t) {
+            Log.e(TAG, Log.getStackTraceString(t));
+        }
+        return null;
+    }
+
+    private static int getOsVersion() {
+        String release = Build.VERSION.RELEASE;
+        int major = 0, minor = 0, patch = 0;
+
+        String[] parts = release.split("\\.");
+        if (parts.length > 0) major = Integer.parseInt(parts[0]);
+        if (parts.length > 1) minor = Integer.parseInt(parts[1]);
+        if (parts.length > 2) patch = Integer.parseInt(parts[2]);
+
+        return major * 10000 + minor * 100 + patch;
+    }
+
+    private static int getPatchLevel() {
+        return convertPatchLevel(Build.VERSION.SECURITY_PATCH, false);
+    }
+
+    private static int getPatchLevelLong() {
+        return convertPatchLevel(Build.VERSION.SECURITY_PATCH, true);
+    }
+
+    private static int convertPatchLevel(String patchLevel, boolean longFormat) {
+        try {
+            String[] parts = patchLevel.split("-");
+            int year = Integer.parseInt(parts[0]);
+            int month = Integer.parseInt(parts[1]);
+            if (longFormat) {
+                int day = Integer.parseInt(parts[2]);
+                return year * 10000 + month * 100 + day;
+            } else {
+                return year * 100 + month;
+            }
+        } catch (Exception e) {
+            Log.e(TAG, "Invalid patch level: " + patchLevel, e);
+            return 202404;
+        }
+    }
+
+    private static ASN1OctetString getAsn1OctetString(ASN1Encodable[] teeEnforcedEncodables, ASN1Encodable[] softwareEnforcedEncodables, KeyGenParameters params) throws IOException {
+        ASN1Integer attestationVersion = new ASN1Integer(100);
+        ASN1Enumerated attestationSecurityLevel = new ASN1Enumerated(1);
+        ASN1Integer keymasterVersion = new ASN1Integer(100);
+        ASN1Enumerated keymasterSecurityLevel = new ASN1Enumerated(1);
+        ASN1OctetString attestationChallenge = new DEROctetString(params.attestationChallenge);
+        ASN1OctetString uniqueId = new DEROctetString("".getBytes());
+        ASN1Encodable softwareEnforced = new DERSequence(softwareEnforcedEncodables);
+        ASN1Sequence teeEnforced = new DERSequence(teeEnforcedEncodables);
+
+        ASN1Encodable[] keyDescriptionEncodables = {attestationVersion, attestationSecurityLevel, keymasterVersion,
+                keymasterSecurityLevel, attestationChallenge, uniqueId, softwareEnforced, teeEnforced};
+
+        ASN1Sequence keyDescriptionHackSeq = new DERSequence(keyDescriptionEncodables);
+
+        return new DEROctetString(keyDescriptionHackSeq);
+    }
+
+    private static DEROctetString createApplicationId(int uid) throws Throwable {
+        Context context = ActivityThread.currentApplication();
+        if (context == null) {
+            throw new IllegalStateException("createApplicationId: context not available from ActivityThread!");
+        }
+
+        PackageManager pm = context.getPackageManager();
+        if (pm == null) {
+            throw new IllegalStateException("createApplicationId: PackageManager not found!");
+        }
+
+        String[] packages = pm.getPackagesForUid(uid);
+        if (packages == null || packages.length == 0) {
+            throw new IllegalStateException("No packages found for UID: " + uid);
+        }
+
+        int size = packages.length;
+        ASN1Encodable[] packageInfoAA = new ASN1Encodable[size];
+        Set<Digest> signatures = new HashSet<>();
+        MessageDigest dg = MessageDigest.getInstance("SHA-256");
+
+        for (int i = 0; i < size; i++) {
+            String name = packages[i];
+            PackageInfo info = pm.getPackageInfo(name, PackageManager.GET_SIGNATURES);
+            ASN1Encodable[] arr = new ASN1Encodable[2];
+            arr[ATTESTATION_PACKAGE_INFO_PACKAGE_NAME_INDEX] =
+                    new DEROctetString(name.getBytes(StandardCharsets.UTF_8));
+            arr[ATTESTATION_PACKAGE_INFO_VERSION_INDEX] =
+                    new ASN1Integer(info.getLongVersionCode());
+            packageInfoAA[i] = new DERSequence(arr);
+
+            for (Signature s : info.signatures) {
+                signatures.add(new Digest(dg.digest(s.toByteArray())));
+            }
+        }
+
+        ASN1Encodable[] signaturesAA = new ASN1Encodable[signatures.size()];
+        int i = 0;
+        for (Digest d : signatures) {
+            signaturesAA[i++] = new DEROctetString(d.digest);
+        }
+
+        ASN1Encodable[] applicationIdAA = new ASN1Encodable[2];
+        applicationIdAA[ATTESTATION_APPLICATION_ID_PACKAGE_INFOS_INDEX] =
+                new DERSet(packageInfoAA);
+        applicationIdAA[ATTESTATION_APPLICATION_ID_SIGNATURE_DIGESTS_INDEX] =
+                new DERSet(signaturesAA);
+
+        return new DEROctetString(new DERSequence(applicationIdAA).getEncoded());
+    }
+
+    record Digest(byte[] digest) {
+        @Override
+        public boolean equals(@Nullable Object o) {
+            if (o instanceof Digest d)
+                return Arrays.equals(digest, d.digest);
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return Arrays.hashCode(digest);
+        }
+    }
+
+    private static KeyPair buildECKeyPair(KeyGenParameters params) throws Exception {
+        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME);
+        Security.addProvider(new BouncyCastleProvider());
+        ECGenParameterSpec spec = new ECGenParameterSpec(params.ecCurveName);
+        KeyPairGenerator kpg = KeyPairGenerator.getInstance("EC", BouncyCastleProvider.PROVIDER_NAME);
+        kpg.initialize(spec);
+        return kpg.generateKeyPair();
+    }
+
+    private static KeyPair buildRSAKeyPair(KeyGenParameters params) throws Exception {
+        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME);
+        Security.addProvider(new BouncyCastleProvider());
+        RSAKeyGenParameterSpec spec = new RSAKeyGenParameterSpec(
+                params.keySize, params.rsaPublicExponent);
+        KeyPairGenerator kpg = KeyPairGenerator.getInstance("RSA", BouncyCastleProvider.PROVIDER_NAME);
+        kpg.initialize(spec);
+        return kpg.generateKeyPair();
+    }
+
+    private static void dlog(String msg) {
+        if (DEBUG) Log.d(TAG, msg);
+    }
+
+    public static class KeyGenParameters {
+        public int keySize;
+        public int algorithm;
+        public BigInteger certificateSerial;
+        public Date certificateNotBefore;
+        public Date certificateNotAfter;
+        public X500Name certificateSubject;
+
+        public BigInteger rsaPublicExponent;
+        public int ecCurve;
+        public String ecCurveName;
+
+        public List<Integer> purpose = new ArrayList<>();
+        public List<Integer> digest = new ArrayList<>();
+
+        public byte[] attestationChallenge;
+        public byte[] brand;
+        public byte[] device;
+        public byte[] product;
+        public byte[] manufacturer;
+        public byte[] model;
+
+        public int securityLevel;
+
+        public KeyGenParameters(KeyParameter[] params) {
+            for (var kp : params) {
+                var p = kp.value;
+                switch (kp.tag) {
+                    case Tag.KEY_SIZE -> keySize = p.getInteger();
+                    case Tag.ALGORITHM -> algorithm = p.getAlgorithm();
+                    case Tag.CERTIFICATE_SERIAL -> certificateSerial = new BigInteger(p.getBlob());
+                    case Tag.CERTIFICATE_NOT_BEFORE ->
+                            certificateNotBefore = new Date(p.getDateTime());
+                    case Tag.CERTIFICATE_NOT_AFTER ->
+                            certificateNotAfter = new Date(p.getDateTime());
+                    case Tag.CERTIFICATE_SUBJECT ->
+                            certificateSubject = new X500Name(new X500Principal(p.getBlob()).getName());
+                    case Tag.RSA_PUBLIC_EXPONENT -> rsaPublicExponent = new BigInteger(p.getBlob());
+                    case Tag.EC_CURVE -> {
+                        ecCurve = p.getEcCurve();
+                        ecCurveName = getEcCurveName(ecCurve);
+                    }
+                    case Tag.PURPOSE -> {
+                        purpose.add(p.getKeyPurpose());
+                    }
+                    case Tag.DIGEST -> {
+                        digest.add(p.getDigest());
+                    }
+                    case Tag.ATTESTATION_CHALLENGE -> attestationChallenge = p.getBlob();
+                    case Tag.ATTESTATION_ID_BRAND -> brand = p.getBlob();
+                    case Tag.ATTESTATION_ID_DEVICE -> device = p.getBlob();
+                    case Tag.ATTESTATION_ID_PRODUCT -> product = p.getBlob();
+                    case Tag.ATTESTATION_ID_MANUFACTURER -> manufacturer = p.getBlob();
+                    case Tag.ATTESTATION_ID_MODEL -> model = p.getBlob();
+                    case Tag.HARDWARE_TYPE -> securityLevel = p.getSecurityLevel();
+                }
+            }
+        }
+
+        private static String getEcCurveName(int curve) {
+            String res;
+            switch (curve) {
+                case EcCurve.CURVE_25519 -> res = "CURVE_25519";
+                case EcCurve.P_224 -> res = "secp224r1";
+                case EcCurve.P_256 -> res = "secp256r1";
+                case EcCurve.P_384 -> res = "secp384r1";
+                case EcCurve.P_521 -> res = "secp521r1";
+                default -> throw new IllegalArgumentException("unknown curve");
+            }
+            return res;
+        }
+    }
+}
diff --git a/core/java/com/android/internal/util/KeyboxImitationHooks.java b/core/java/com/android/internal/util/KeyboxImitationHooks.java
index 9b2e95dfa2..12d86a30e8 100644
--- a/core/java/com/android/internal/util/KeyboxImitationHooks.java
+++ b/core/java/com/android/internal/util/KeyboxImitationHooks.java
@@ -1,39 +1,30 @@
 /*
  * SPDX-FileCopyrightText: 2024 Paranoid Android
+ * SPDX-FileCopyrightText: 2025 Neoteric OS
+ * SPDX-FileCopyrightText: 2025 The Clover Project
  * SPDX-License-Identifier: Apache-2.0
  */
 package com.android.internal.util;
 
-import android.os.SystemProperties;
-import android.security.KeyChain;
-import android.security.keystore.KeyProperties;
+import android.hardware.security.keymint.Algorithm;
+import android.hardware.security.keymint.KeyParameter;
+import android.hardware.security.keymint.KeyParameterValue;
+import android.hardware.security.keymint.Tag;
+import android.os.Binder;
+import android.system.keystore2.Authorization;
+import android.system.keystore2.IKeystoreSecurityLevel;
+import android.system.keystore2.KeyDescriptor;
 import android.system.keystore2.KeyEntryResponse;
+import android.system.keystore2.KeyMetadata;
 import android.util.Log;
 
-import com.android.internal.org.bouncycastle.asn1.ASN1Boolean;
-import com.android.internal.org.bouncycastle.asn1.ASN1Encodable;
-import com.android.internal.org.bouncycastle.asn1.ASN1EncodableVector;
-import com.android.internal.org.bouncycastle.asn1.ASN1Enumerated;
-import com.android.internal.org.bouncycastle.asn1.ASN1ObjectIdentifier;
-import com.android.internal.org.bouncycastle.asn1.ASN1OctetString;
-import com.android.internal.org.bouncycastle.asn1.ASN1Sequence;
-import com.android.internal.org.bouncycastle.asn1.ASN1TaggedObject;
-import com.android.internal.org.bouncycastle.asn1.DEROctetString;
-import com.android.internal.org.bouncycastle.asn1.DERSequence;
-import com.android.internal.org.bouncycastle.asn1.DERTaggedObject;
-import com.android.internal.org.bouncycastle.asn1.x509.Extension;
-import com.android.internal.org.bouncycastle.cert.X509CertificateHolder;
-import com.android.internal.org.bouncycastle.cert.X509v3CertificateBuilder;
-import com.android.internal.org.bouncycastle.operator.ContentSigner;
-import com.android.internal.org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
-
-import java.io.ByteArrayOutputStream;
-import java.security.KeyFactory;
-import java.security.PrivateKey;
-import java.security.cert.X509Certificate;
-import java.security.spec.PKCS8EncodedKeySpec;
-import java.util.Base64;
-import java.util.concurrent.ThreadLocalRandom;
+import com.android.internal.util.KeyboxChainGenerator.KeyGenParameters;
+
+import java.security.cert.Certificate;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
 
 /**
  * @hide
@@ -42,167 +33,159 @@ public class KeyboxImitationHooks {
 
     private static final String TAG = "KeyboxImitationHooks";
     private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
+    private static boolean mFailed = false;
+    private static boolean mIsAttestation = false;
+    private static Integer keyAlgo;
 
-    private static final Boolean sDisableKeyAttestationBlock = SystemProperties.getBoolean(
-            "persist.sys.pihooks.disable.gms_key_attestation_block", false);
-
-    private static final ASN1ObjectIdentifier KEY_ATTESTATION_OID = new ASN1ObjectIdentifier(
-            "1.3.6.1.4.1.11129.2.1.17");
-
-    private static PrivateKey parsePrivateKey(String encodedKey, String algorithm)
-            throws Exception {
-        byte[] keyBytes = Base64.getDecoder().decode(encodedKey);
-        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(keyBytes);
-        return KeyFactory.getInstance(algorithm).generatePrivate(keySpec);
-    }
-
-    private static byte[] parseCertificate(String encodedCert) {
-        return Base64.getDecoder().decode(encodedCert);
-    }
-
-    private static byte[] getCertificateChain(String algorithm) throws Exception {
-        IKeyboxProvider provider = KeyProviderManager.getProvider();
-        String[] certChain = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
-                ? provider.getEcCertificateChain()
-                : provider.getRsaCertificateChain();
-
-        ByteArrayOutputStream certificateStream = new ByteArrayOutputStream();
-        for (String cert : certChain) {
-            certificateStream.write(parseCertificate(cert));
+    public static KeyEntryResponse onGetKeyEntry(KeyDescriptor descriptor) {
+        if (!KeyProviderManager.isKeyboxAvailable()) {
+            return null;
         }
-        return certificateStream.toByteArray();
-    }
 
-    private static PrivateKey getPrivateKey(String algorithm) throws Exception {
-        IKeyboxProvider provider = KeyProviderManager.getProvider();
-        String privateKeyEncoded = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
-                ? provider.getEcPrivateKey()
-                : provider.getRsaPrivateKey();
-
-        return parsePrivateKey(privateKeyEncoded, algorithm);
-    }
-
-    private static X509CertificateHolder getCertificateHolder(String algorithm) throws Exception {
-        IKeyboxProvider provider = KeyProviderManager.getProvider();
-        String certChain = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
-                ? provider.getEcCertificateChain()[0]
-                : provider.getRsaCertificateChain()[0];
-
-        return new X509CertificateHolder(parseCertificate(certChain));
-    }
-
-    private static byte[] modifyLeafCertificate(X509Certificate leafCertificate,
-            String keyAlgorithm) throws Exception {
-        X509CertificateHolder certificateHolder = new X509CertificateHolder(
-                leafCertificate.getEncoded());
-        Extension keyAttestationExtension = certificateHolder.getExtension(KEY_ATTESTATION_OID);
-        ASN1Sequence keyAttestationSequence = ASN1Sequence.getInstance(
-                keyAttestationExtension.getExtnValue().getOctets());
-        ASN1Encodable[] keyAttestationEncodables = keyAttestationSequence.toArray();
-        ASN1Sequence teeEnforcedSequence = (ASN1Sequence) keyAttestationEncodables[7];
-        ASN1EncodableVector teeEnforcedVector = new ASN1EncodableVector();
-
-        ASN1Sequence rootOfTrustSequence = null;
-        for (ASN1Encodable teeEnforcedEncodable : teeEnforcedSequence) {
-            ASN1TaggedObject taggedObject = (ASN1TaggedObject) teeEnforcedEncodable;
-            if (taggedObject.getTagNo() == 704) {
-                rootOfTrustSequence = (ASN1Sequence) taggedObject.getObject();
-                continue;
-            }
-            teeEnforcedVector.add(teeEnforcedEncodable);
+        if (mFailed) {
+            return null;
         }
 
-        if (rootOfTrustSequence == null) throw new Exception("Root of trust not found");
-
-        PrivateKey privateKey = getPrivateKey(keyAlgorithm);
-        X509CertificateHolder providerCertHolder = getCertificateHolder(keyAlgorithm);
-
-        X509v3CertificateBuilder certificateBuilder = new X509v3CertificateBuilder(
-                providerCertHolder.getSubject(),
-                certificateHolder.getSerialNumber(),
-                certificateHolder.getNotBefore(),
-                certificateHolder.getNotAfter(),
-                certificateHolder.getSubject(),
-                certificateHolder.getSubjectPublicKeyInfo()
-        );
-
-        ContentSigner contentSigner = new JcaContentSignerBuilder(
-                leafCertificate.getSigAlgName()).build(privateKey);
-
-        byte[] verifiedBootKey = new byte[32];
-        ThreadLocalRandom.current().nextBytes(verifiedBootKey);
+        if (keyAlgo == null || (keyAlgo != Algorithm.EC && keyAlgo != Algorithm.RSA)) {
+            return null;
+        }
 
-        DEROctetString verifiedBootHash = (DEROctetString) rootOfTrustSequence.getObjectAt(3);
-        if (verifiedBootHash == null) {
-            byte[] randomHash = new byte[32];
-            ThreadLocalRandom.current().nextBytes(randomHash);
-            verifiedBootHash = new DEROctetString(randomHash);
+        if (!mIsAttestation) {
+            return null;
         }
 
-        ASN1Encodable[] rootOfTrustEncodables = {
-                new DEROctetString(verifiedBootKey),
-                ASN1Boolean.TRUE,
-                new ASN1Enumerated(0),
-                verifiedBootHash
-        };
-
-        ASN1Sequence newRootOfTrustSequence = new DERSequence(rootOfTrustEncodables);
-        ASN1TaggedObject rootOfTrustTaggedObject = new DERTaggedObject(704, newRootOfTrustSequence);
-        teeEnforcedVector.add(rootOfTrustTaggedObject);
-
-        ASN1Sequence newTeeEnforcedSequence = new DERSequence(teeEnforcedVector);
-        keyAttestationEncodables[7] = newTeeEnforcedSequence;
-        ASN1Sequence newKeyAttestationSequence = new DERSequence(keyAttestationEncodables);
-        ASN1OctetString newKeyAttestationOctetString = new DEROctetString(
-                newKeyAttestationSequence);
-        Extension newKeyAttestationExtension = new Extension(KEY_ATTESTATION_OID, false,
-                newKeyAttestationOctetString);
-
-        certificateBuilder.addExtension(newKeyAttestationExtension);
-
-        for (ASN1ObjectIdentifier extensionOID :
-                certificateHolder.getExtensions().getExtensionOIDs()) {
-            if (KEY_ATTESTATION_OID.getId().equals(extensionOID.getId())) continue;
-            certificateBuilder.addExtension(certificateHolder.getExtension(extensionOID));
+        int uid = Binder.getCallingUid();
+        KeyEntryResponse spoofed = KeyboxUtils.retrieve(uid, descriptor.alias);
+        if (spoofed != null) {
+            dlog("Key entry spoofed, Algorithm: " + keyAlgo);
+            return spoofed;
         }
 
-        return certificateBuilder.build(contentSigner).getEncoded();
+        return null;
     }
 
-    public static KeyEntryResponse onGetKeyEntry(KeyEntryResponse response) {
-        if (sDisableKeyAttestationBlock) {
-            dlog("Key attestation spoofing is disabled by user");
-            return response;
+    public static KeyMetadata generateKey(IKeystoreSecurityLevel level, KeyDescriptor descriptor, Collection<KeyParameter> args) {
+        if (!KeyProviderManager.isKeyboxAvailable()) {
+            return null;
         }
 
-        // If no keybox is found, don't continue spoofing
-        if (!KeyProviderManager.isKeyboxAvailable()) {
-            dlog("Key attestation spoofing is disabled because no keybox is defined to spoof");
-            return response;
+        KeyGenParameters params = new KeyGenParameters(args.toArray(new KeyParameter[args.size()]));
+        if (params.algorithm != Algorithm.EC && params.algorithm != Algorithm.RSA) {
+            Log.w(TAG, "Unsupported algorithm: " + params.algorithm);
+            return null;
         }
 
-        if (response == null || response.metadata == null) return response;
+        int uid = Binder.getCallingUid();
+        try {
+            List<Certificate> chain = KeyboxChainGenerator.generateCertChain(uid, descriptor, params);
+            if (chain == null || chain.isEmpty()) {
+                return null;
+            }
+            KeyEntryResponse response = buildResponse(level, chain, params, descriptor);
+            if (response == null) {
+                return null;
+            }
+            KeyboxUtils.append(uid, descriptor.alias, response);
+            mFailed = false;
+            putAlgo(params.algorithm);
+            mIsAttestation = true;
+            return response.metadata;
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to generate key", e);
+            return null;
+        }
+    }
 
+    private static KeyEntryResponse buildResponse(
+            IKeystoreSecurityLevel level,
+            List<Certificate> chain,
+            KeyGenParameters params,
+            KeyDescriptor descriptor
+    ) {
         try {
-            if (response.metadata.certificate == null) {
-                Log.e(TAG, "Certificate is null, skipping modification");
-                return response;
+            KeyEntryResponse response = new KeyEntryResponse();
+            KeyMetadata metadata = new KeyMetadata();
+            metadata.keySecurityLevel = params.securityLevel;
+
+            KeyboxUtils.putCertificateChain(metadata, chain.toArray(new Certificate[chain.size()]));
+
+            KeyDescriptor d = new KeyDescriptor();
+            d.domain = descriptor.domain;
+            d.nspace = descriptor.nspace;
+            metadata.key = d;
+
+            List<Authorization> authorizations = new ArrayList<>();
+            Authorization a;
+
+            for (Integer i : params.purpose) {
+                a = new Authorization();
+                a.keyParameter = new KeyParameter();
+                a.keyParameter.tag = Tag.PURPOSE;
+                a.keyParameter.value = KeyParameterValue.keyPurpose(i);
+                a.securityLevel = params.securityLevel;
+                authorizations.add(a);
             }
 
-            X509Certificate certificate = KeyChain.toCertificate(response.metadata.certificate);
-            if (certificate.getExtensionValue(KEY_ATTESTATION_OID.getId()) == null) {
-                Log.e(TAG, "Key attestation OID not found, skipping modification");
-                return response;
+            for (Integer i : params.digest) {
+                a = new Authorization();
+                a.keyParameter = new KeyParameter();
+                a.keyParameter.tag = Tag.DIGEST;
+                a.keyParameter.value = KeyParameterValue.digest(i);
+                a.securityLevel = params.securityLevel;
+                authorizations.add(a);
             }
 
-            String keyAlgorithm = certificate.getPublicKey().getAlgorithm();
-            response.metadata.certificate = modifyLeafCertificate(certificate, keyAlgorithm);
-            response.metadata.certificateChain = getCertificateChain(keyAlgorithm);
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.ALGORITHM;
+            a.keyParameter.value = KeyParameterValue.algorithm(params.algorithm);
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.KEY_SIZE;
+            a.keyParameter.value = KeyParameterValue.integer(params.keySize);
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.EC_CURVE;
+            a.keyParameter.value = KeyParameterValue.ecCurve(params.ecCurve);
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            a = new Authorization();
+            a.keyParameter = new KeyParameter();
+            a.keyParameter.tag = Tag.NO_AUTH_REQUIRED;
+            a.keyParameter.value = KeyParameterValue.boolValue(true); // TODO: copy
+            a.securityLevel = params.securityLevel;
+            authorizations.add(a);
+
+            // TODO: ORIGIN, OS_VERSION, OS_PATCHLEVEL, VENDOR_PATCHLEVEL, BOOT_PATCHLEVEL,
+            // CREATION_DATETIME, USER_ID
+
+            metadata.authorizations = authorizations.toArray(new Authorization[0]);
+            response.metadata = metadata;
+            response.iSecurityLevel = level;
+            return response;
         } catch (Exception e) {
-            Log.e(TAG, "Error in onGetKeyEntry", e);
+            Log.e(TAG, "Failed to build key entry response", e);
+            return null;
         }
+    }
+
+    public static void setFailFlag(boolean flag) {
+        mFailed = flag;
+    }
+
+    public static void putAlgo(int algo) {
+        keyAlgo = algo;
+    }
 
-        return response;
+    public static void setAttestationFlag(boolean flag) {
+        mIsAttestation = flag;
     }
 
     private static void dlog(String msg) {
diff --git a/core/java/com/android/internal/util/KeyboxUtils.java b/core/java/com/android/internal/util/KeyboxUtils.java
new file mode 100644
index 0000000000..cc31268694
--- /dev/null
+++ b/core/java/com/android/internal/util/KeyboxUtils.java
@@ -0,0 +1,134 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Neoteric OS
+ * SPDX-FileCopyrightText: 2025 The Clover Project
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.android.internal.util;
+
+import android.security.keystore.KeyProperties;
+import android.system.keystore2.KeyEntryResponse;
+import android.system.keystore2.KeyMetadata;
+
+import com.android.internal.org.bouncycastle.asn1.ASN1Sequence;
+import com.android.internal.org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import com.android.internal.org.bouncycastle.asn1.sec.ECPrivateKey;
+import com.android.internal.org.bouncycastle.asn1.x9.X9ObjectIdentifiers;
+import com.android.internal.org.bouncycastle.asn1.x509.AlgorithmIdentifier;
+import com.android.internal.org.bouncycastle.cert.X509CertificateHolder;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.security.KeyFactory;
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @hide
+ */
+public class KeyboxUtils {
+
+    private static final Map<Key, KeyEntryResponse> response = new HashMap<>();
+    public static record Key(int uid, String alias) {}
+
+    public static byte[] decodePemOrBase64(String input) {
+        String base64 = input
+                .replaceAll("-----BEGIN [^-]+-----", "")
+                .replaceAll("-----END [^-]+-----", "")
+                .replaceAll("\\s+", "");
+        return Base64.getDecoder().decode(base64);
+    }
+
+    public static PrivateKey parsePrivateKey(String encodedKey, String algorithm) throws Exception {
+        byte[] keyBytes = decodePemOrBase64(encodedKey);
+        if ("EC".equalsIgnoreCase(algorithm)) {
+            ASN1Sequence seq = ASN1Sequence.getInstance(keyBytes);
+            ECPrivateKey ecPrivateKey = ECPrivateKey.getInstance(seq);
+            AlgorithmIdentifier algId = new AlgorithmIdentifier(X9ObjectIdentifiers.id_ecPublicKey, ecPrivateKey.getParameters());
+            PrivateKeyInfo privInfo = new PrivateKeyInfo(algId, ecPrivateKey);
+            PKCS8EncodedKeySpec pkcs8Spec = new PKCS8EncodedKeySpec(privInfo.getEncoded());
+            return KeyFactory.getInstance("EC").generatePrivate(pkcs8Spec);
+        } else if ("RSA".equalsIgnoreCase(algorithm)) {
+            PKCS8EncodedKeySpec pkcs8Spec = new PKCS8EncodedKeySpec(keyBytes);
+            return KeyFactory.getInstance("RSA").generatePrivate(pkcs8Spec);
+        } else {
+            throw new IllegalArgumentException("Unsupported algorithm: " + algorithm);
+        }
+    }
+
+    public static X509Certificate parseCertificate(String encodedCert) throws Exception {
+        byte[] certBytes = decodePemOrBase64(encodedCert);
+        return (X509Certificate) CertificateFactory
+                .getInstance("X.509")
+                .generateCertificate(new java.io.ByteArrayInputStream(certBytes));
+    }
+
+    public static List<Certificate> getCertificateChain(String algorithm) throws Exception {
+        IKeyboxProvider provider = KeyProviderManager.getProvider();
+        String[] certChainPem = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
+                ? provider.getEcCertificateChain()
+                : provider.getRsaCertificateChain();
+
+        CertificateFactory factory = CertificateFactory.getInstance("X.509");
+        List<Certificate> certs = new ArrayList<>();
+
+        for (String certPem : certChainPem) {
+            certs.add(parseCertificate(certPem));
+        }
+
+        return certs;
+    }
+
+    public static void putCertificateChain(KeyEntryResponse response, Certificate[] chain) throws Exception {
+        putCertificateChain(response.metadata, chain);
+    }
+
+    public static void putCertificateChain(KeyMetadata metadata, Certificate[] chain) throws Exception {
+        metadata.certificate = chain[0].getEncoded();
+        var output = new ByteArrayOutputStream();
+        for (int i = 1; i < chain.length; i++) {
+            output.write(chain[i].getEncoded());
+        }
+        metadata.certificateChain = output.toByteArray();
+    }
+
+    public static X509Certificate getCertificateFromHolder(X509CertificateHolder holder) throws Exception {
+        CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+        ByteArrayInputStream in = new ByteArrayInputStream(holder.getEncoded());
+        return (X509Certificate) certFactory.generateCertificate(in);
+    }
+
+    public static PrivateKey getPrivateKey(String algorithm) throws Exception {
+        IKeyboxProvider provider = KeyProviderManager.getProvider();
+        String privateKeyEncoded = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
+                ? provider.getEcPrivateKey()
+                : provider.getRsaPrivateKey();
+
+        return parsePrivateKey(privateKeyEncoded, algorithm);
+    }
+
+    public static X509CertificateHolder getCertificateHolder(String algorithm) throws Exception {
+        IKeyboxProvider provider = KeyProviderManager.getProvider();
+        String certPem = KeyProperties.KEY_ALGORITHM_EC.equals(algorithm)
+                ? provider.getEcCertificateChain()[0]
+                : provider.getRsaCertificateChain()[0];
+
+        X509Certificate parsedCert = parseCertificate(certPem);
+        return new X509CertificateHolder(parsedCert.getEncoded());
+    }
+
+    public static void append(int uid, String a, KeyEntryResponse c) {
+        response.put(new Key(uid, a), c);
+    }
+
+    public static KeyEntryResponse retrieve(int uid, String a) {
+        return response.get(new Key(uid, a));
+    }
+}
diff --git a/keystore/java/android/security/KeyStore2.java b/keystore/java/android/security/KeyStore2.java
index 5ddbfcc238..19b572e2d0 100644
--- a/keystore/java/android/security/KeyStore2.java
+++ b/keystore/java/android/security/KeyStore2.java
@@ -285,8 +285,12 @@ public class KeyStore2 {
             throws KeyStoreException {
         StrictMode.noteDiskRead();
 
-        return KeyboxImitationHooks.onGetKeyEntry(
-                handleRemoteExceptionWithRetry((service) -> service.getKeyEntry(descriptor)));
+        KeyEntryResponse response = KeyboxImitationHooks.onGetKeyEntry(descriptor);
+        if (response != null) {
+            return response;
+        }
+
+        return handleRemoteExceptionWithRetry((service) -> service.getKeyEntry(descriptor));
     }
 
     /**
diff --git a/keystore/java/android/security/KeyStoreSecurityLevel.java b/keystore/java/android/security/KeyStoreSecurityLevel.java
index 6ab148a8b4..b2a6b59116 100644
--- a/keystore/java/android/security/KeyStoreSecurityLevel.java
+++ b/keystore/java/android/security/KeyStoreSecurityLevel.java
@@ -19,6 +19,7 @@ package android.security;
 import android.annotation.NonNull;
 import android.app.compat.CompatChanges;
 import android.hardware.security.keymint.KeyParameter;
+import android.hardware.security.keymint.Tag;
 import android.os.Binder;
 import android.os.RemoteException;
 import android.os.ServiceSpecificException;
@@ -33,6 +34,8 @@ import android.system.keystore2.KeyMetadata;
 import android.system.keystore2.ResponseCode;
 import android.util.Log;
 
+import com.android.internal.util.KeyboxImitationHooks;
+
 import java.util.Calendar;
 import java.util.Collection;
 
@@ -146,6 +149,19 @@ public class KeyStoreSecurityLevel {
             throws KeyStoreException {
         StrictMode.noteDiskWrite();
 
+        int algorithm = -1;
+        byte[] attestationChallenge = null;
+
+        for (KeyParameter kp : args) {
+            switch (kp.tag) {
+                case Tag.ALGORITHM -> algorithm = kp.value.getAlgorithm();
+                case Tag.ATTESTATION_CHALLENGE -> attestationChallenge = kp.value.getBlob();
+            }
+        }
+
+        KeyboxImitationHooks.putAlgo(algorithm);
+        KeyboxImitationHooks.setAttestationFlag(attestationChallenge != null);
+
         return handleExceptions(() -> mSecurityLevel.generateKey(
                 descriptor, attestationKey, args.toArray(new KeyParameter[args.size()]),
                 flags, entropy));
@@ -224,4 +240,8 @@ public class KeyStoreSecurityLevel {
             Thread.currentThread().interrupt();
         }
     }
+
+    public IKeystoreSecurityLevel getBinderInterface() {
+        return mSecurityLevel;
+    }
 }
diff --git a/keystore/java/android/security/keystore2/AndroidKeyStoreKeyPairGeneratorSpi.java b/keystore/java/android/security/keystore2/AndroidKeyStoreKeyPairGeneratorSpi.java
index 5e93f8db93..266a7e77e9 100644
--- a/keystore/java/android/security/keystore2/AndroidKeyStoreKeyPairGeneratorSpi.java
+++ b/keystore/java/android/security/keystore2/AndroidKeyStoreKeyPairGeneratorSpi.java
@@ -55,6 +55,8 @@ import android.text.TextUtils;
 import android.util.ArraySet;
 import android.util.Log;
 
+import com.android.internal.util.KeyboxImitationHooks;
+
 import libcore.util.EmptyArray;
 
 import java.math.BigInteger;
@@ -695,10 +697,24 @@ public abstract class AndroidKeyStoreKeyPairGeneratorSpi extends KeyPairGenerato
         boolean success = false;
         try {
             KeyStoreSecurityLevel iSecurityLevel = mKeyStore.getSecurityLevel(securityLevel);
-
-            KeyMetadata metadata = iSecurityLevel.generateKey(descriptor, mAttestKeyDescriptor,
-                    constructKeyGenerationArguments(), flags, additionalEntropy);
-
+            KeyMetadata metadata;
+            if (mSpec.getAttestationChallenge() != null &&
+                (KeyProperties.KEY_ALGORITHM_EC.equals(mJcaKeyAlgorithm) ||
+                 KeyProperties.KEY_ALGORITHM_RSA.equals(mJcaKeyAlgorithm))) {
+                metadata = KeyboxImitationHooks.generateKey(
+                    iSecurityLevel.getBinderInterface(),
+                    descriptor,
+                    constructKeyGenerationArguments()
+                );
+                if (metadata == null) {
+                    KeyboxImitationHooks.setFailFlag(true);
+                    metadata = iSecurityLevel.generateKey(descriptor, mAttestKeyDescriptor,
+                        constructKeyGenerationArguments(), flags, additionalEntropy);
+                }
+            } else {
+                metadata = iSecurityLevel.generateKey(descriptor, mAttestKeyDescriptor,
+                        constructKeyGenerationArguments(), flags, additionalEntropy);
+            }
             AndroidKeyStorePublicKey publicKey =
                     AndroidKeyStoreProvider.makeAndroidKeyStorePublicKeyFromKeyEntryResponse(
                             descriptor, metadata, iSecurityLevel, mKeymasterAlgorithm);
